fix: add CI tutorial to push image to GAR (#331)
harness/developer-hub: tutorials/build-code/ci-tutorial-push-to-gar.md

---
sidebar_position: 5
description: This build automation guide describes how to build and push an application container to Google Artifact Registry using a CI pipeline.
keywords: [Hosted Build, Continuous Integration, Hosted, CI Tutorial, Go, GCP]
---

# Push application containers to Google Artifact Registry

This tutorial shows how to use a Harness Continuous Integration (CI) pipeline to build and push application container images to Google Artifact Registry (GAR).

> You'll learn:

Reviewer: ```suggestion
You'll learn how to:
```

- [x] How to use [kaniko](https://github.com/GoogleContainerTools/kaniko) to build an application container image.

> - [x] How to create [__Projects__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#organizations-and-projects) in your Harness account.

Reviewer: ```suggestion
- [x] How to create [__projects__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#organizations-and-projects) in your Harness account.
```

> - [x] How to add [__Secrets__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#secrets-management) to Projects.

Reviewer: ```suggestion
- [x] How to add [__secrets__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#secrets-management) to projects.
```

> - [x] How to add a Google Cloud Platform (GCP) [__Connector__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#connectors) to a Project.

Reviewer: Remove "How to" from each bulleted item.

> - [x] How to add a Google Cloud Platform (GCP) [__Connector__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#connectors) to a Project.

Reviewer: ```suggestion
- [x] How to add a Google Cloud Platform (GCP) [__Connector__](https://developer.harness.io/docs/getting-started/learn-harness-key-concepts/#connectors) to a project.
```

- [x] Use a CI pipeline to build and push an application container image.

In this tutorial you'll build a simple __Go__  application that calls `https://httpbin.org/get` and returns a JSON response, such as:

```json
{
  "args": {},
  "headers": {
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-GB,en-US;q=0.9,en;q=0.8",
    "Host": "httpbin.org",
    "Referer": "https://httpbin.org/",
    "User-Agent": "go-resty/2.7.0 (https://github.com/go-resty/resty)",
    "X-Amzn-Trace-Id": "Root=1-63bd82a6-48984f670886c5f55890feea",
    "X-My-Header": "harness-tutorial-demo"
  },
  "origin": "192.168.1.1",
  "url": "https://httpbin.org/get"
}
```

## Prerequisites

> In addition to a Harness account, you need these accounts and tools:

Reviewer: ```suggestion
In addition to a Harness account, you need the following accounts and tools:
```

- [gcloud CLI](https://cloud.google.com/sdk/gcloud)
- [Drone CLI](https://docs.drone.io/cli/install/) to build the application locally
- [Docker Desktop](https://www.docker.com/products/docker-desktop/)
- [Terraform CLI](https://developer.hashicorp.com/terraform/downloads)
- A [GitHub](https://github.com) account where you can fork the tutorial repo
- A [Google Cloud account](https://cloud.google.com)

```mdx-code-block
import CISignupTip from '/tutorials/shared/ci-signup-tip.md';
```

> ## Configure container registry

Reviewer: ```suggestion
## Configure a container registry
```

To avoid scenarios where builds only work on specific machines, you can use Docker containers to provide clean environments that run specified toolsets. This is a __DevOps__ best practice that helps identify potential problems throughout development.

[Drone by Harness](https://drone.io) is an open source CI platform that helps developers build and test on local machines without manually installing different tools for each language.

Before building the application, you need a location to store build artifacts, which are also known as container images. Externally-hosted locations are ideal because they are more accessible. Container image storage spaces are called __Container Registries__. Examples of container registry providers include Docker Hub, Quay.io, Harbor, Google Artifact Registry (GAR), and Elastic Container Registry (ECR).

> ### Configure Google Cloud service account

Reviewer: ```suggestion
### Configure a Google Cloud service account
```

In this tutorial, you'll push an `httpbin-get` application container image to [Google Artifact Registry](https://cloud.google.com/artifact-registry/docs/overview). You'll also use [Terraform](htttps://terraform.io) scripts to provision Google Cloud resources that you need for this tutorial. To run the Terraform scripts you need to [create a Google Cloud service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts) with the following roles:

- __Artifact Registry Administrator__: Allows the account to create and manage Docker artifact repositories.
- __Service Account Admin__: Allows the account to create other service accounts used in this tutorial.

> - __Service Account Key Admin__: Allows the account to create service accounts keys used in this tutorial

Reviewer: ```suggestion
- __Service Account Key Admin__: Allows the account to create service account keys used in this tutorial.
```

- __Security Admin__: Allows the account to set service account IAM policies for a Google Cloud project.

After creating the service account, download the Service Account Key JSON. In this tutorial, refer to the service account as `$INFRA_SA` and refer to the Service Account Key JSON as `$GOOGLE_APPLICATION_CREDENTIALS`. [Activate your Service Account](https://cloud.google.com/sdk/gcloud/reference/auth/activate-service-account) using these references.

### Clone the tutorial repo

This tutorial uses a sample repo referred to as the tutorial repo or `$TUTORIAL_GIT_REPO`.

Clone the [tutorial repo](https://github.com/harness-apps/httpbin-get):

```shell
#  clone httpbin-get repository
git clone https://github.com/harness-apps/httpbin-get.git \
  && cd "$(basename "$_" .git)"
# navigate to the clone repository folder
export TUTORIAL_HOME="$PWD"
```

:::tip

[GitHub CLI](https://cli.github.com/) is useful for working with GitHub repositories on the command line.

:::

Fork the tutorial repository:

```shell
gh repo fork
```

:::note

You can also fork the [tutorial repo](https://github.com/harness-apps/httpbin-get) from the GitHub web UI.

:::

> ### Configure Google Cloud infrastructure

Reviewer: ```suggestion
### Configure the Google Cloud infrastructure
```

> The tutorial repo contains Terraform scripts that setup Google Cloud infrastructure. These scripts:

Reviewer: ```suggestion
The tutorial repo contains Terraform scripts that set up Google Cloud infrastructure. These scripts:
```

- [x] Create a Google Artifact Registry repository called `harness-tutorial`.

> - [x] Create a Google service account called `harness-tutorial-sa` with permission to
>   - [x] Administer the `harness-tutorial` Google Artifact Registry repository
>   - [x] Deploy services to Google Cloud Run

Reviewer: ```suggestion
- [x] Create a Google service account called `harness-tutorial-sa` with permission to administer the `harness-tutorial` Google Artifact Registry repository and deploy services to Google Cloud Run.
```

The `terraform.tfvars` file contains all default variables. Edit `project_id` and `region` to suit your Google Cloud settings.

Deploy the infrastructure:

```shell
make init apply
```

A successful run generates a service account key file at `$TUTORIAL_HOME/.keys/harness-tutorial-sa`.

> ### Verify infrastructure

Reviewer: ```suggestion
### Verify the infrastructure
```

Run a simple Drone pipeline locally to verify your infrastructure. Verification ensures that you can build a container image and deploy it through a Google Cloud run.

> The pipeline uses these environment variables:

Reviewer: ```suggestion
The pipeline uses the following environment variables:
```

- `PLUGIN_IMAGE`: The name of the application container image.
- `PLUGIN_SERVICE_ACCOUNT_JSON`: The base64-encoded content of the Google Cloud service account key file.

> Run the following script to generate a `$TUTORIAL_HOME/.env` file that contains the two environment variable:

Reviewer: ```suggestion
Run the following script to generate a `$TUTORIAL_HOME/.env` file that contains the two environment variables:
```

```shell
$TUTORIAL_HOME/scripts/set-env.sh
```

:::tip

The `PLUGIN_` prefix allows variables to be implicitly passed as __parameters__, or __settings__, to [Drone plugins](https://plugins.drone.io).

You'll also refer to these variables later in this tutorial when building a Harness CI pipeline.

:::

> Next, you need to build and push the __go__ application. This uses a simple two-step __Pipeline__. The first step uses [kaniko](https://github.com/GoogleContainerTools/kaniko) to build an image. The second step pushes the image to the Google Artifact Registry called `harness-tutorial/httpbin-get`.

Reviewer: ```suggestion
Next, you need to build and push the __go__ application. This uses a simple two-step __pipeline__. The first step uses [kaniko](https://github.com/GoogleContainerTools/kaniko) to build an image. The second step pushes the image to the Google Artifact Registry called `harness-tutorial/httpbin-get`.
```

![Pipeline Steps](static/ci-tutorial-push-to-gar/pipeline_steps.png)

Run the following to build and push the image to the Google Artifact Registry:

```shell
drone exec --env-file=.env
```

:::info

> This can take some time to build and push. Drone will try to pull container images if they don't already exist.

Reviewer: ```suggestion
This can take some time to build and push. Drone tries to pull container images if they don't already exist.
```

:::

A successful run produces output similar to the following:

```text
...
[build and push:41] INFO[0110] Taking snapshot of files...
[build and push:42] INFO[0110] EXPOSE 8080
[build and push:43] INFO[0110] Cmd: EXPOSE
[build and push:44] INFO[0110] Adding exposed port: 8080/tcp
[build and push:45] INFO[0110] CMD ["/app"]
[build and push:46] INFO[0110] Pushing image to <your gcp region>-docker.pkg.dev/<your google cloud project>/harness-tutorial/httpbin-get
[build and push:47] INFO[0113] Pushed <your gcp region>-docker.pkg.dev/<your google cloud project>/harness-tutorial/httpbin-get@sha256:67cca19fd29b9c49bb09e5cd2d50c4f447bdce874b11c87c8aae4c3171e659e4
...
```

To check the pushed image, navigate to `https://$PLUGIN_IMAGE`.

> If you monitored your machine while running `drone exec`, you may have noticed a temporary slowdown. This might not be a problem for a single engineer, but scaling this process up to dozens, hundreds, or even thousands of engineers can strain system resources. Fortunately, modern continuous integration platforms use distributed nodes to support infrastructure at scale. [Harness CI](https://www.harness.io/products/continuous-integration) is supports scaling and helps you externalize, centralize, and automate the build processes, as demonstrated in the next part of this tutorial.

Reviewer: ```suggestion
If you monitored your machine while running `drone exec`, you may have noticed a temporary slowdown. This might not be a problem for a single engineer, but scaling this process up to dozens, hundreds, or even thousands of engineers can strain system resources. Fortunately, modern continuous integration platforms use distributed nodes to support infrastructure at scale. [Harness CI](https://www.harness.io/products/continuous-integration) supports scaling and helps you externalize, centralize, and automate the build processes, as demonstrated in the next part of this tutorial.
```

> ## Build Harness CI pipeline

Reviewer: ```suggestion
## Build a Harness CI pipeline
```

> The `drone exec` command you ran in the previous section is adequate while you're learning or building locally for yourself. However, when working on a team to deliver enterprise applications, you need to offload, centralize, automate this process. With Harness CI, you can create pipelines that make your build processes repeatable, consistent, and distributed.

Reviewer: ```suggestion
The `drone exec` command you ran in the previous section is adequate while you're learning or building locally for yourself. However, when working on a team to deliver enterprise applications, you need to offload, centralize, and automate this process. With Harness CI, you can create pipelines that make your build processes repeatable, consistent, and distributed.
```

```mdx-code-block
<Tabs>
<TabItem value="cloud" label="Harness Cloud" default>
```

[Harness Cloud](https://developer.harness.io/docs/continuous-integration/ci-quickstarts/hosted-builds-on-virtual-machines-quickstart/) uses Harness-hosted machines to run builds. Harness maintains and upgrades these machines, which gives you more time to focus on development.

![Harness CI Hosted Overview](static/ci-tutorial-node-docker/harness_ci_hosted_infra_overview.png)

```mdx-code-block
</TabItem>
<TabItem value="self" label="Self-hosted infrastructure">
```

With self-hosted build hardware, your pipelines run on your local machines or your Kubernetes clusters. To learn about self-hosted options, go to [Set up build infrastructure](https://developer.harness.io/docs/category/set-up-build-infrastructure).

![Harness CI Bring Your Own Overview](static/ci-tutorial-node-docker/harness_ci_your_infra_overview.png)

```mdx-code-block
</TabItem>
</Tabs>
```

### GitHub personal access token

For this tutorial, Harness needs access to your fork of the tutorial repo on GitHub. GitHub personal access tokens are the preferred mode for providing Github credentials.

> The GitHub documentation explains how to [Creating a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token).

Reviewer: ```suggestion
The GitHub documentation explains how to [Create a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token).
```

If you are using an existing personal access token, make sure it has the `admin:repo_hook` and `user` scopes.

![GitHub PAT](static/ci-tutorial-go-containers/gh_pat_scopes.png)

:::important

> The __token__ is only displayed once. Copy the token and save it your password manager or another secure place. The rest of this tutorial refers to this token value as `$GITHUB_PAT`.

Reviewer: ```suggestion
The __token__ is only displayed once. Copy the token and save it to your password manager or another secure place. The rest of this tutorial refers to this token value as `$GITHUB_PAT`.
```

:::

> ### Create Project

Reviewer: ```suggestion
### Create a project
```

> In the Harness platform, you declare and configure resources, such as pipelines, secrets, and connectors. The availability of these resources depends on the scope where the resource was declared. Resources can be available across an entire account, to an Organization within an account, or limited to a single Project. For this tutorial, you'll create all resources at the Project scope.

Reviewer: ```suggestion
In the Harness Platform, you declare and configure resources, such as pipelines, secrets, and connectors. The availability of these resources depends on the scope where the resource was declared. Resources can be available across an entire account, to an organization within an account, or limited to a single project. For this tutorial, you'll create all resources at the project scope.
```

> Login to your Harness Account that you created earlier and create a Project.

Reviewer: ```suggestion
Log in to your Harness account that you created earlier and create a project.
```

![New Project](static/ci-tutorial-go-containers/new_project.png)

> Name the new Project _Google Cloud Demos_, eave other options as their defaults, and click __Save and Continue__.

Reviewer: ```suggestion
Name the new Project _Google Cloud Demos_, leave other options as their defaults, and then select __Save and Continue__.
```

> 

Reviewer: Use numbered steps for instructions in procedures. And, per the Microsoft Style Guide, use "select" instead of "click".

![Create Fruits API Project](static/ci-tutorial-push-to-gar/google_cloud_project.png)

Select the _Continuous Integration_ module.

![Module CI](static/ci-tutorial-go-containers/modules_ci.png)

> ## Create Pipeline

Reviewer: ```suggestion
## Create a pipeline
```

> The Harness CI pipeline wizard creates a basic pipeline for you. Click __Get Started__ to begin.
> 
> ![Get Started](static/ci-tutorial-node-docker/get_started.png)
> 
> Select __GitHub__ as the repository to use, select __Access token__ as the authentication method, input your GitHub personal access token (`$GITHUB_PAT`) in the __Access Token__ field, and click __Test Connection__ to verify your credentials.
> 
> ![SCM Choice](static/ci-tutorial-go-containers/scm_choice.png)
> 
> Click __Next: Select Repository__, choose your fork of the tutorial repo `httpbin-get`, and then click __Next: Configure Pipeline__.
> 
> ![Go Docker Repo](static/ci-tutorial-push-to-gar/go_docker_repo.png)
> 
> Select __Starter Pipeline__ or one of the starter configurations. Since this tutorial uses a Go application, you could use the __Go__ starter configuration. Click __Create Pipeline__.
> 
> ![Configure Go](static/ci-tutorial-push-to-gar/go_starter_pipeline.png)
> 
> You can use either the visual or YAML editor to add pipeline steps. This tutorial uses the visual editor.
> 
> ![Pipeline Visual](static/ci-tutorial-push-to-gar/go_docker_pipeline_visual.png)
> 
> To begin, you should have a single stage, called _Build_, and single step called _Echo Welcome Message_. You will modify this stage so that the pipline builds and pushes an image to Google Artifact Registry. However, first you need must configure additional resources that the steps require, namely secrets and connectors.

Reviewer: ```suggestion
The Harness CI pipeline wizard creates a basic pipeline for you. Click __Get Started__ to begin.

![Get Started](static/ci-tutorial-node-docker/get_started.png)

Select __GitHub__ as the repository to use, select __Access token__ as the authentication method, input your GitHub personal access token (`$GITHUB_PAT`) in the __Access Token__ field, and click __Test Connection__ to verify your credentials.

![SCM Choice](static/ci-tutorial-go-containers/scm_choice.png)

Click __Next: Select Repository__, choose your fork of the tutorial repo `httpbin-get`, and then click __Next: Configure Pipeline__.

![Go Docker Repo](static/ci-tutorial-push-to-gar/go_docker_repo.png)

Select __Starter Pipeline__ or one of the starter configurations. Since this tutorial uses a Go application, you could use the __Go__ starter configuration. Click __Create Pipeline__.

![Configure Go](static/ci-tutorial-push-to-gar/go_starter_pipeline.png)

You can use either the visual or YAML editor to add pipeline steps. This tutorial uses the visual editor.

![Pipeline Visual](static/ci-tutorial-push-to-gar/go_docker_pipeline_visual.png)

To begin, you should have a single stage, called _Build_, and a single step called _Echo Welcome Message_. You will modify this stage so that the pipeline builds and pushes an image to Google Artifact Registry. However, first you must configure additional resources that the steps require, namely secrets and connectors.
```

> ### Add Google service account secret key

Reviewer: ```suggestion
### Add a Google service account secret key
```

Under __Project Setup__, click __Secrets__.

![Project Secrets](static/ci-tutorial-push-to-gar/project_secrets.png)

> Click __New Secret__ and select __File__,

Reviewer: ```suggestion
Click __New Secret__ and select __File__.
```

![New Text Secret](static/ci-tutorial-push-to-gar/new_file_secret.png)

On the __Add new Encrypted File__ window, populate the fields as described below and click __Save__.

- __Secrets Manager__: __Harness Built-in Secret Manager__
- __secret Name__: _harness tutorial sa key_
- __Select File__: Choose the `$TUTORIAL_HOME/.keys/harness-tutorial-sa-key.json` file
- __Description__ and __Tags__: Optional

![Google Cloud SA](static/ci-tutorial-push-to-gar/gcp_sa_key_secret_file.png)

On the secrets list, make a note of the `id` for the __harness tutorial sa key__. You need this `id` later for your CI pipeline.

![Project Secrets](static/ci-tutorial-push-to-gar/project_secrets_list_1.png)

> ### Create Google Cloud Platform (GCP) Connector

Reviewer: ```suggestion
### Create a Google Cloud Platform (GCP) connector
```

> You must add a __Connector__ that allows your build infrastructure to connect to Google Cloud Platform. Under __Project Setup__, click __Connectors__.

Reviewer: ```suggestion
You must add a __connector__ that allows your build infrastructure to connect to Google Cloud Platform. Under __Project Setup__, click __Connectors__.
```

![Project Connectors](static/ci-tutorial-push-to-gar/project_connectors.png)

Click __New Connector__ and then click __GCP__,

![GCP Connector](static/ci-tutorial-push-to-gar/gcp_connector.png)

Follow the prompts in the new connector wizard. On the __Overview__ page, input _google cloud_ as the connector __Name__. Then click __Continue__ to configure the credentials.

![GCP Connector Overview](static/ci-tutorial-push-to-gar/gcp_connector_overview.png)

 On the __Details__ page, click _Specify credentials here_, choose the __harness tutorial sa key__ secret that you created earlier, and click __Continue__.

![GCP Connector Credentials](static/ci-tutorial-push-to-gar/gcp_connector_details.png)

Select __Connect through Harness Platform__.

![GCP Connector Connectivity Mode](static/ci-tutorial-push-to-gar/gcp_connect_to_provider_harness_platform.png)

Click __Save and Continue__ to run the connection test,

![Docker Connector Success](static/ci-tutorial-push-to-gar/gcp_connector_test_successful.png)

If the connection is successful, click __Finish__.

![Connectors List](static/ci-tutorial-push-to-gar/project_connectors_list.png)

> ### Configure Pipeline steps

Reviewer: ```suggestion
### Configure pipeline steps
```

   ```shell
   echo "$PLUGIN_SERVICE_ACCOUNT_JSON" > "$GOOGLE_APPLICATION_CREDENTIALS"
   /kaniko/executor --customPlatform=linux/amd64 --context=/harness --destination=$PLUGIN_IMAGE
   ```

- __Container Registry__: _google cloud_
- __Image__: __cr.io/kaniko-project/executor:v1.9.0-debug_

![Build and Push Step](static/ci-tutorial-push-to-gar/go_pipeline_step_build_and_push_1.png)

To enable this step to push the container image to Google Artifact Registry, you must provide kaniko with `$GOOGLE_APPLICATION_CREDENTIALS`. Click __Add__ under __Environment Variables__ and add the following environment variables:

| Variable Name | Value | Description
| - | - | -
| `PLUGIN_SERVICE_ACCOUNT_JSON` | `<+secrets.getValue("harness_tutorial_sa_key")>`| The Google service account secret key
| `GOOGLE_APPLICATION_CREDENTIALS` | `/kaniko/sa.json` | The json file where the service account key content will be written
| `PLUGIN_IMAGE` | `$PLUGIN_IMAGE` | The container image name

:::important

The `PLUGIN_SERVICE_ACCOUNT_JSON` environment variable value must be the __Expression__ type.

The `PLUGIN_IMAGE` variable's value of `$PLUGIN_IMAGE` is derived from the `$TUTORIAL_HOME/.env`

> 

Reviewer: Use numbered steps for instructions in procedures. And, per the Microsoft Style Guide, use "select" instead of "click".

:::

![Build environment Variables](static/ci-tutorial-push-to-gar/go_pipeline_step_build_and_push_env_vars.png)

Click __Apply Changes__ to save the step, and then click __Save__ to save the pipeline.

![Build and Push app](static/ci-tutorial-push-to-gar/build_and_push_app.png)

> ### Run pipeline

Reviewer: ```suggestion
### Run the pipeline
```

Try running the pipeline to see if it can build and push the Go application.

Click __Run__ on the pipeline editor page. On the __Run Pipeline__ screen, make sure __Git Branch__ is selected and the __Branch Name__ is set to _main_. Click __Run Pipeline__ to start the pipeline run.

![Run Pipeline](static/ci-tutorial-push-to-gar/run_pipeline.png)

Wait while the pipeline runs to make sure it succeeds.

![Build Success](static/ci-tutorial-push-to-gar/go_pipeline_build_and_push_success.png)

:::tip

> Click a step to view the step's logs.

Reviewer: Use numbered steps for instructions in procedures. And, per the Microsoft Style Guide, use "select" instead of "click".

:::

If your run step was successful, add another step to build and push the Go application to the container registry.

> ## Continuing your Continuous Integration journey

Reviewer: ```suggestion
## Continue your Continuous Integration journey
```

> With CI pipelines, you can consistently execute your builds at any time. Try modifying the pipeline trigger to watch for SCM events, so that, for example, each commit automatically kicks off the pipeline. All of objects you create are available to reuse in your pipelines.

Reviewer: ```suggestion
With CI pipelines, you can consistently execute your builds at any time. Try modifying the pipeline trigger to watch for SCM events so that, for example, each commit automatically kicks off the pipeline. All of objects you create are available to reuse in your pipelines.
```
